remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,549 @@
+  uint64_t offset;
+  for (auto &operand : I.operands()) {
+    if (exclude.count(operand)) {
+      in_exclude_set = true;
+    } else {
+      auto ptr = offset_map.find(operand);
+      if (ptr == offset_map.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        if (!in_offset_map) {
+          offset = ptr->second;
+          in_offset_map = true;
+        } else {
+          if (ptr->second != offset) {
+            // bail if the offsets don't match
+            return VisitResult::Error;
+          }
+        }
+      }
+    }
+  }
+  if (in_offset_map && in_exclude_set) {
+    // fail if some operands are excluded and others are state offsets
+    return VisitResult::Error;
+  } else if (in_offset_map) {
+    LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+    offset_map.insert({&I, offset});
+  } else if (in_exclude_set) {
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    exclude.insert(&I);
+  }
+  return VisitResult::Progress;
+}
+
+// Return the scope of the given instruction.
+llvm::MDNode *GetScopeFromInst(llvm::Instruction &I) {
+  llvm::AAMDNodes N;
+  I.getAAMetadata(N);
+  return N.Scope;
+}
+
+// Remove all dead stores.
+void RemoveDeadStores(const std::unordered_set<llvm::Instruction *> &dead_stores) {
+  for (auto *store : dead_stores) {
+    store->eraseFromParent();
+  }
+}
+
+LiveSetBlockVisitor::LiveSetBlockVisitor(const ScopeMap &scope_to_offset_)
+  : scope_to_offset(scope_to_offset_),
+    curr_wl(),
+    next_wl(),
+    block_map(),
+    to_remove() {}
+
+// Get every terminating basic block from the function `func`.
+void LiveSetBlockVisitor::AddFunction(llvm::Function &func) {
+  for (auto &block : func) {
+    // The machines rose from the ashes of the nuclear fire....
+    if (block.getTerminator()) {
+      curr_wl.push_back(&block);
+      // add block to map with initial entry set
+      block_map.insert({&block, LiveSet()});
+    }
+  }
+}
+
+// Visit the basic blocks in the worklist and update the block_map.
+void LiveSetBlockVisitor::Visit(void) {
+  //bool progress = true;
+  auto currlive = new LiveSet;
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty()) {
+    LOG(INFO) << "LSBV: " << curr_wl.size() << " blocks in worklist";
+    //progress = false;
+    for (auto block : curr_wl) {
+      auto newlive = VisitBlock(block, currlive);
+      if (*newlive == *currlive) {
+        // converged
+        goto end;
+      }
+      LOG(INFO) << "LSBV: " << newlive->count() << " live slots";
+      currlive = newlive;
+    }
+    // update curr_wl to the next set of insts
+    curr_wl.swap(next_wl);
+    // reset upcoming insts
+    next_wl.clear();
+  }
+end:
+  LOG(INFO) << "Found " << to_remove.size() << " dead stores to remove";
+  RemoveDeadStores(to_remove);
+}
+
+LiveSet *LiveSetBlockVisitor::VisitBlock(llvm::BasicBlock *B, LiveSet *init) {
+  // Set the exit value based on the initial liveset unioned with the successors
```

Comment:
I suggest you start with something like this:
```c++
LiveSet live;
for (auto inst_it = B->rbegin(); inst_it != B->rend(); ++inst_it) {
  auto inst = &*inst_it;
  if (llvm::isa<llvm::ReturnInst>(inst) || llvm::isa<llvm::UnreachableInst>(inst)) {
    // mark all live
  } else if (auto br_inst = llvm::dyn_cast<llvm::BrInst>(inst)) {
    // update live set from successors
  } else if (auto call_inst = llvm::dyn_cast<llvm::CallInst>(inst)) {
    // mark all live if the called function type is the same as a lifted function type
  } else if (auto invoke_inst = llvm::dyn_cast<llvm::InvokeInst>(inst)) {
    // treated like a call
  } else if (llvm::isa<llvm::LoadInst>(inst)) {
    ...
  } else if (llvm::isa<llvm::StoreInst>(inst)) {
    ...
  }
}
```

The loop updates `live`. At the end of the function, you have something like:
```c++
auto &old_live_on_entry = block_map[B];
if (old_live_on_entry != live) {
  old_live_on_entry = live;
  return true;
} else {
  return false;
}
```